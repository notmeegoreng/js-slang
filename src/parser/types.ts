import type { Program } from 'estree'

import type { Context, Node, Chapter, Variant, SourceError } from '../types'

export type { ParserOptions as BabelOptions } from '@babel/parser'
export type { Options as AcornOptions } from 'acorn'

export interface Parser<TOptions> {
  parse(
    programStr: string,
    context: Context,
    options?: Partial<TOptions>,
    throwOnError?: boolean
  ): Program | null
  validate(ast: Program, context: Context, throwOnError?: boolean): boolean
}
export interface Rule<T extends Node> {
  /**
   * Name of the rule
   */
  name: string

  /**
   * Test snippets to test the behaviour of the rule. Providing no test snippets
   * means that this rule will not be tested when running unit tests.\
   * First element of the tuple is the code to test. Set the second element to `undefined`
   * if the snippet should not throw an error. Otherwise set it to the `explain()` value
   * of the error.
   */
  testSnippets?: [code: string, expected: string | undefined][]

  /**
   * Disable this rule for this chapter (inclusive) and above
   */
  disableFromChapter?: Chapter
  disableForVariants?: Variant[]
  checkers: {
    [name: string]: (node: T, ancestors: Node[]) => SourceError[]
  }
}
