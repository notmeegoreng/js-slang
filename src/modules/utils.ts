import * as _ from 'lodash'

import { RecursivePartial } from '../types'
import { ImportOptions } from './moduleTypes'
import { defaultAnalysisOptions } from './preprocessor/analyzer'

const exportDefaultStr = 'export default'
export function removeExportDefault(text: string) {
  if (text.startsWith(exportDefaultStr)) {
    text = text.substring(exportDefaultStr.length).trim()
  }

  return text
}

export function mergeImportOptions(src?: RecursivePartial<ImportOptions>): ImportOptions {
  const baseOptions = _.cloneDeep(defaultAnalysisOptions)
  return _.merge(baseOptions, src as any)
}

export const isSourceModule = (path: string) => !path.startsWith('.') && !path.startsWith('/')
