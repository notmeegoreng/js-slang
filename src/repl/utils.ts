import { Option } from '@commander-js/extra-typings'
import { Chapter, Variant } from '../types'

export function chapterParser(str: string): Chapter {
  let foundChapter: string | undefined

  if (/^[0-9]$/.test(str)) {
    // Chapter is fully numeric
    const value = parseInt(str)
    foundChapter = Object.keys(Chapter).find(chapterName => Chapter[chapterName] === value)
  } else {
    foundChapter = str
  }

  if (foundChapter === undefined) {
    throw new Error(`Invalid chapter value: ${str}`)
  }

  return Chapter[foundChapter]
}

export const chapterOption = new Option('--chapter <chapter>')
  .default(Chapter.SOURCE_4)
  .argParser(chapterParser)

export const variantOption = new Option('--variant <variant>')
  .default(Variant.DEFAULT)
  .choices(Object.values(Variant))
