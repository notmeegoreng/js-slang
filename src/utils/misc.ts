import { RuntimeSourceError } from '../errors/runtimeSourceError'

export class PromiseTimeoutError extends RuntimeSourceError {}

export const timeoutPromise = <T>(promise: Promise<T>, timeout: number) =>
  new Promise<T>((resolve, reject) => {
    const timeoutid = setTimeout(() => reject(new PromiseTimeoutError()), timeout)

    promise
      .then(res => {
        clearTimeout(timeoutid)
        resolve(res)
      })
      .catch(e => {
        clearTimeout(timeoutid)
        reject(e)
      })
  })

/**
 * Run the mapping function over the items, filtering out any items that
 * that the mapping function returned `undefined` for
 */
export function mapAndFilter<T, U>(items: T[], mapper: (input: T) => U | undefined) {
  return items.reduce((res, item) => {
    const newItem = mapper(item)
    if (newItem !== undefined) return [...res, newItem]
    return res
  }, [] as U[])
}

export function mapObject<T extends Record<any, any>, U>(
  obj: T,
  mapper: (key: keyof T, value: T[keyof T]) => U
): Record<keyof T, U> {
  return objectEntries(obj).reduce(
    (res, [key, value]) => ({
      ...res,
      [key]: mapper(key, value)
    }),
    {} as Record<keyof T, U>
  )
}

/**
 * Type safe `Object.keys`
 */
export function objectKeys<T extends string | number | symbol>(obj: Record<T, any>): T[] {
  return Object.keys(obj) as T[]
}

/**
 * Type safe `Object.values`
 */
export function objectValues<T>(obj: Record<any, T>) {
  return Object.values(obj) as T[]
}

type DeconstructRecord<T extends Record<any, any>> = {
  [K in keyof T]: [K, T[K]]
}

type EntriesOfRecord<T extends Record<any, any>> = DeconstructRecord<T>[keyof T]

export function objectEntries<T extends Record<any, any>>(obj: T): EntriesOfRecord<T>[] {
  return Object.entries(obj)
}
