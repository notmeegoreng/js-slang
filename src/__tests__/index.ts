import { Chapter, type Value } from '../types'
import { expectResult } from '../utils/testing'

const toString = (x: Value) => '' + x

test('Arrow function definition returns itself', () => {
  return expectResult('() => 42;').toMatchInlineSnapshot(`[Function]`)
})

test('Builtins hide their implementation when toString', () => {
  return expectResult('toString(pair);', {
    chapter: Chapter.SOURCE_2,
    native: true,
    testBuiltins: { toString }
  }).toMatchInlineSnapshot(`
            "function pair(left, right) {
            	[implementation hidden]
            }"
          `)
})
