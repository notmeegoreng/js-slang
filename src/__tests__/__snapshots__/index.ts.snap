// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Arrow function definition returns itself: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "() => 42;",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": [Function],
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins hide their implementation when toString: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "toString(pair);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "function pair(left, right) {
	[implementation hidden]
}",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;
